src/solvers/td3.ml

```diff
@@ -145,48 +296,117 @@ module WP =
+                 Otherwise recursive solve might prematurely abort it. *)
+              if tracing then trace "sol2" "solve still unstable %a\n" S.Var.pretty_trace x;
+              (solve[@tailcall]) x Widen changed
+            ) else (
+              if abort && HM.mem destab_front x then (
+                if tracing then trace "sol2" "front remove %a\n" S.Var.pretty_trace x;
+                HM.remove destab_front x;
+                if tracing then trace "sol2" "not pushing front from %a\n" S.Var.pretty_trace x;
+                (* don't push front here *)
+                HM.remove destab_infl x
+              );
+              if term && phase = Widen && HM.mem wpoint x then ( (* TODO: or use wp? *)
+                if tracing then trace "sol2" "solve switching to narrow %a\n" S.Var.pretty_trace x;
+                if tracing then trace "sol2" "stable remove %a\n" S.Var.pretty_trace x;
+                HM.remove stable x;
+                HM.remove superstable x;
+                let reuse_eq = if not aborted then
+                    Some new_eq
+                  else
+                    None (* TODO: for some reason cannot reuse aborted rhs, see 34-localwn_restart/02-hybrid *)
+                in
+                (solve[@tailcall]) ?reuse_eq ~abortable:false x Narrow changed
+              ) else if not space && (not term || phase = Narrow) then ( (* this makes e.g. nested loops precise, ex. tests/regression/34-localization/01-nested.c - if we do not remove wpoint, the inner loop head will stay a wpoint and widen the outer loop variable. *)
+                if tracing then trace "sol2" "solve removing wpoint %a (%b)\n" S.Var.pretty_trace x (HM.mem wpoint x);
+                HM.remove wpoint x;
+                changed
+              )
+              else
+                changed
+            )
           )
         )
+        else if HM.mem called x then
+          changed || HM.mem called_changed x
+        else
+          changed
       and eq x get set =
         if tracing then trace "sol2" "eq %a\n" S.Var.pretty_trace x;
         eval_rhs_event x;
         match S.system x with
         | None -> S.Dom.bot ()
         | Some f -> f get set
-      and simple_solve l x y =
+      and simple_solve l x y: S.d * bool =
         if tracing then trace "sol2" "simple_solve %a (rhs: %b)\n" S.Var.pretty_trace y (S.system y <> None);
-        if S.system y = None then (init y; HM.find rho y) else
-        if HM.mem rho y || not space then (solve y Widen; HM.find rho y) else
-        if HM.mem called y then (init y; HM.remove l y; HM.find rho y) else
+        if S.system y = None then (init y; (HM.find rho y, true (* TODO: ??? *))) else
+        if HM.mem rho y || not space then (let changed = solve y Widen false in (HM.find rho y, changed)) else
+        if abort then failwith "space abort unimplemented" else
+        if HM.mem called y then (init y; HM.remove l y; (HM.find rho y, true (* TODO: ??? *))) else
```

Comment:
Is this `TODO: ???` resolved

Reply:
For reference, as mentioned in GobCon these are about using `space` (with caching) and `abort` at the same time, but in that case the previous line already raises an exception.

I wasn't sure how the combination of caching and aborting would work because the implementation of caching notably differs from the one presented in the appendix of the TD3 paper: there are numerous extra cases here.